ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,1352 @@
+  offset_trampoline: u32,
+  offset_original: u32,
+  stack_allocated: u32,
+}
+
+#[cfg_attr(
+  not(all(target_aarch = "x86_64", target_family = "windows")),
+  allow(dead_code)
+)]
+impl Win64 {
+  // > Integer valued arguments in the leftmost four positions are passed in left-to-right order in RCX, RDX, R8, and R9
+  // > [...]
+  // > Any floating-point and double-precision arguments in the first four parameters are passed in XMM0 - XMM3, depending on position
+  const REGISTERS: u32 = 4;
+
+  fn new() -> Self {
+    Self {
+      assembler: dynasmrt::x64::Assembler::new().unwrap(),
+      params: 0,
+      // trampoline caller's return address + trampoline's shadow space
+      offset_trampoline: 8 + 32,
+      offset_original: 8 + 32,
+      stack_allocated: 0,
+    }
+  }
+
+  fn compile(sym: &Symbol) -> Trampoline {
+    let mut compiler = Self::new();
+
+    let can_tailcall = !compiler.must_cast_return_value(sym.result_type);
+    if !can_tailcall {
+      compiler.allocate_stack(&sym.parameter_types);
+    }
+
+    for argument in &sym.parameter_types {
+      compiler.move_left(argument)
+    }
+    if !compiler.is_recv_overriden() {
+      // the receiver object should never be expected. Avoid its unexpected or deliverated leak
+      compiler.zero_first_arg();
+    }
+
+    if !can_tailcall {
+      compiler.call(sym.ptr.as_ptr());
+      if compiler.must_cast_return_value(sym.result_type) {
+        compiler.cast_return_value(sym.result_type);
+      }
+      compiler.deallocate_stack();
+      compiler.ret();
+    } else {
+      compiler.tailcall(sym.ptr.as_ptr());
+    }
+
+    Trampoline(compiler.finalize())
+  }
+
+  fn move_left(&mut self, arg: &NativeType) {
+    match arg {
+      NativeType::F32 | NativeType::F64 => self.move_arg(F),
+      NativeType::U8 | NativeType::I8 => self.move_arg(I(B)),
+      NativeType::U16 | NativeType::I16 => self.move_arg(I(W)),
+      NativeType::U32 | NativeType::I32 | NativeType::Void => {
+        self.move_arg(I(DW))
+      }
+      NativeType::U64
+      | NativeType::USize
+      | NativeType::Function
+      | NativeType::Pointer
+      | NativeType::I64
+      | NativeType::ISize => self.move_arg(I(QW)),
+    }
+  }
+
+  fn move_arg(&mut self, param: Param) {
+    let param_i = self.params + 1;
+    self.params = param_i;
+
+    // move each argument one position to the left. The first argument in the stack moves to the last register (r9).
+    // If the FFI function is called with a new stack frame, the arguments remaining in the stack are copied to the new stack frame.
+    // Otherwise, they are copied 8 bytes lower in the same frame
+    match (param_i, param) {
+      (1, I(B | W | DW)) => dynasm!(self.assembler; .arch x64; mov ecx, edx),
+      (1, I(QW)) => dynasm!(self.assembler; .arch x64; mov rcx, rdx),
+      (1, F) => {
+        // Use movaps for doubles as well, benefits of smaller encoding outweight those of using the correct instruction for the type (movapd)
```

Comment:
Maybe add a note about zeroing out `ecx` here due to it not getting erased by an integer parameter.

Reply:
I've added a call to `zero_first_arg()` instead, is this explicit enough?

Reply:
Yeah, absolutely great! <3